x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java

```diff
@@ -0,0 +1,145 @@
+    public static JoinType FULL = CoreJoinType.FULL;
+    public static JoinType CROSS = CoreJoinType.CROSS;
+
+    /**
+     * The predefined core join types. Implements as enum for easy comparison and serialization.
+     */
+    private enum CoreJoinType implements JoinType {
+        INNER(1, "INNER"),
+        LEFT(2, "LEFT OUTER"),
+        RIGHT(3, "RIGHT OUTER"),
+        FULL(4, "FULL OUTER"),
+        CROSS(5, "CROSS");
+
+        private final String name;
+        private final byte id;
+
+        CoreJoinType(int id, String name) {
+            this.id = (byte) id;
+            this.name = name;
+        }
+
+        @Override
+        public String joinName() {
+            return name;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeByte(id);
+        }
+    }
+
+    /**
+     * Join type for the USING clause - shorthand for defining an equi-join (equality join meaning the condition checks if columns across
+     * each side of the join are equal).
+     * One important difference is that the USING clause returns the join column only once, at the beginning of the result set.
+     */
+    public static class UsingJoinType implements JoinType {
+        private final List<Attribute> columns;
+        private final JoinType coreJoin;
+
+        public UsingJoinType(JoinType coreJoin, List<Attribute> columns) {
+            this.columns = columns;
+            this.coreJoin = coreJoin;
+        }
+
+        @Override
+        public String joinName() {
+            return coreJoin.joinName() + " USING " + columns.toString();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new IllegalArgumentException("USING join type should not be serialized due to being rewritten");
+        }
+
+        public JoinType coreJoin() {
+            return coreJoin;
+        }
+
+        public List<Attribute> columns() {
+            return columns;
+        }
+
+        @Override
+        public boolean resolved() {
+            return Resolvables.resolved(columns);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(columns, coreJoin);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            UsingJoinType that = (UsingJoinType) o;
+            return Objects.equals(columns, that.columns) && coreJoin == that.coreJoin;
+        }
+
+        @Override
+        public String toString() {
+            return joinName();
+        }
+    }
+
+    /**
+     * Private class so it doesn't get used yet it is defined to showcase why the join type was defined as an interface instead of a simpler
+     * enum.
+     */
+    private abstract static class NaturalJoinType implements JoinType {
```

Comment:
As `USING` establishes a natural  join, why not use/expand this "type"/class, instead of the `UsingJoinType`? Will we ever use them both?

Reply:
Using and Natural are similar but different:
- in Using the keys are specified by the user. Using supports Inner, Left, right and full outer join.
- Natural - the keys are detected at runtime. Only supports inner join.

Hence why these are modelled differently.